// amend/dist/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AgentConfig = exports.CoreEngine = exports.Registry = exports.ValidationError = exports.NetworkError = exports.ScopeError = void 0;
const scopeError_js_1 = require("./errors/scopeError.js");
Object.defineProperty(exports, "ScopeError", { enumerable: true, get: function () { return scopeError_js_1.ScopeError; } });
const networkError_js_1 = require("./errors/networkError.js");
Object.defineProperty(exports, "NetworkError", { enumerable: true, get: function () { return networkError_js_1.NetworkError; } });
const validationError_js_1 = require("./errors/validationError.js");
Object.defineProperty(exports, "ValidationError", { enumerable: true, get: function () { return validationError_js_1.ValidationError; } });
const Registry_js_1 = require("./registry/Registry.js");
const CoreEngine_js_1 = require("./registry/CoreEngine.js");
const AgentConfig_js_1 = require("./ai/AgentConfig.js");
const registry = Registry_js_1.Registry.getInstance();
const engine = CoreEngine_js_1.CoreEngine;
const ai = AgentConfig_js_1.AgentConfig.getInstance();
const amend = {
    ScopeError: scopeError_js_1.ScopeError,
    NetworkError: networkError_js_1.NetworkError,
    ValidationError: validationError_js_1.ValidationError,
    engine: {
        encode: (...args) => CoreEngine_js_1.CoreEngine.encode(...args),
        decode: (key) => CoreEngine_js_1.CoreEngine.decode(key),
        verify: (key) => CoreEngine_js_1.CoreEngine.verify(key),
        errorClassFrom: (key) => CoreEngine_js_1.CoreEngine.errorClassFrom(key),
        locationFrom: (key) => CoreEngine_js_1.CoreEngine.locationFrom(key),
        messageFrom: (key) => CoreEngine_js_1.CoreEngine.messageFrom(key),
    },
    registry: {
        getAll: () => registry.getAll(),
        find: (key) => registry.find(key),
        clear: () => registry.clear(),
        status: () => registry.status(),
        exportTo: (filePath, format) => registry.exportTo(filePath, format),
    },
    ai: {
        enable: (opts) => ai.enable(opts.scope),
        disable: (opts) => ai.disable(opts.scope),
        hints: {
            set: (key, hint) => ai.setHint(key, hint),
            get: (key) => ai.getHint(key),
            clear: (key) => ai.clearHint(key),
        },
    },
};
exports.default = amend;
var Registry_js_2 = require("./registry/Registry.js");
Object.defineProperty(exports, "Registry", { enumerable: true, get: function () { return Registry_js_2.Registry; } });
var CoreEngine_js_2 = require("./registry/CoreEngine.js");
Object.defineProperty(exports, "CoreEngine", { enumerable: true, get: function () { return CoreEngine_js_2.CoreEngine; } });
var AgentConfig_js_2 = require("./ai/AgentConfig.js");
Object.defineProperty(exports, "AgentConfig", { enumerable: true, get: function () { return AgentConfig_js_2.AgentConfig; } });
