// amend/src/errors/AmendError.ts
import { encode as engineEncode } from '../registry/CoreEngine.js';
import { Registry } from '../registry/Registry.js';

export interface AmendErrorOptions {
  context?:  string;
  metadata?: Record<string, unknown>[];
}

export class AmendError extends Error {
  readonly key:       string;
  readonly timestamp: string;
  readonly context?:  string;
  readonly metadata:  Record<string, unknown>[];

  constructor(message: string, errorClass: string, options?: AmendErrorOptions) {
    super(message);
    this.name      = errorClass;
    this.timestamp = new Date().toISOString();

    // ── Extract throw site from stack ──────────────────────────
    const frames  = this.stack?.split('\n') ?? [];
    let filePath  = 'unknown';
    let line      = 0;
    let className = '';
    let funcName  = 'unknown';

    const cleanPath = (p: string) => {
      let normalized = p.replace(/\\/g, '/');
      const cwdPosix = process.cwd().replace(/\\/g, '/');
      const cwdUrl = 'file:///' + cwdPosix;
      if (normalized.startsWith(cwdUrl)) normalized = normalized.replace(cwdUrl + '/', '');
      else if (normalized.startsWith(cwdPosix)) normalized = normalized.replace(cwdPosix + '/', '');
      return normalized;
    };

    for (let i = 2; i < frames.length; i++) {
      const frame = frames[i];
      if (frame.includes('node_modules/amend') || (frame.includes('AmendError') && !frame.includes('.test.'))) continue;

      const m = frame.match(/at (?:(.+?)\.)?(.+?) \((.+?):(\d+):\d+\)/);
      if (m) {
        className = m[1] ?? '';
        funcName  = m[2];
        filePath  = cleanPath(m[3]);
        line = parseInt(m[4], 10);
        break;
      }

      const sm = frame.match(/at (.+?):(\d+):\d+/);
      if (sm) {
        filePath = cleanPath(sm[1]);
        line = parseInt(sm[2], 10);
        break;
      }
    }

    // ── Build self-decoding key via CoreEngine ─────────────────
    this.key      = engineEncode(filePath, line, className, funcName, message, errorClass);
    this.context  = options?.context;
    this.metadata = options?.metadata ?? [];

    // ── Register lean record ───────────────────────────────────
    Registry.getInstance().record(this);
  }
}
