// amend/dist/esm/registry/Registry.js
import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';
import { decode as engineDecode } from './CoreEngine.js';
// ── Helpers ───────────────────────────────────────────────────────────
function safeDecodeKey(key) {
    try {
        return engineDecode(key);
    }
    catch {
        return {};
    }
}
function mergeToRecord(raw) {
    const d = safeDecodeKey(raw.key);
    return {
        key: raw.key,
        errorClass: d.errorClass ?? '',
        message: d.message ?? '',
        filePath: d.filePath ?? '',
        line: d.line ?? 0,
        funcName: d.funcName ?? '',
        timestamp: raw.timestamp,
        count: 1,
        context: raw.context,
        metadata: raw.metadata ?? [],
    };
}
// ── Registry ──────────────────────────────────────────────────────────
export class Registry {
    static instance;
    dir;
    filePath;
    constructor() {
        this.dir = path.join(os.homedir(), '.amend');
        this.filePath = path.join(this.dir, 'errors.log');
        this.ensureDir();
    }
    static getInstance() {
        if (!Registry.instance)
            Registry.instance = new Registry();
        return Registry.instance;
    }
    ensureDir() {
        if (!fs.existsSync(this.dir))
            fs.mkdirSync(this.dir, { recursive: true });
    }
    // ── Write ──────────────────────────────────────────────────────
    record(error) {
        const entry = {
            key: error.key,
            timestamp: error.timestamp,
            ...(error.context ? { context: error.context } : {}),
            ...(error.metadata?.length ? { metadata: error.metadata } : {}),
        };
        fs.appendFileSync(this.filePath, JSON.stringify(entry) + '\n', 'utf8');
    }
    // ── Read ───────────────────────────────────────────────────────
    /**
     * Read all raw lines and consolidate by key:
     *   count    = number of occurrences
     *   metadata = all metadata arrays merged in order
     *   timestamp = most recent occurrence
     */
    getAll() {
        if (!fs.existsSync(this.filePath))
            return [];
        const content = fs.readFileSync(this.filePath, 'utf8').trim();
        if (!content)
            return [];
        const index = new Map();
        for (const rawLine of content.split('\n')) {
            if (!rawLine)
                continue;
            let raw;
            try {
                raw = JSON.parse(rawLine);
            }
            catch {
                continue;
            }
            if (!raw.key)
                continue;
            const existing = index.get(raw.key);
            if (existing) {
                existing.count++;
                existing.timestamp = raw.timestamp; // keep latest
                if (raw.metadata?.length)
                    existing.metadata.push(...raw.metadata);
            }
            else {
                index.set(raw.key, mergeToRecord(raw));
            }
        }
        return [...index.values()];
    }
    find(key) {
        return this.getAll().filter((r) => r.key === key || r.key.startsWith(key));
    }
    clear() {
        if (fs.existsSync(this.filePath))
            fs.writeFileSync(this.filePath, '', 'utf8');
    }
    status() {
        const all = this.getAll();
        return { total: all.length, recent: all.slice(-5) };
    }
    exportTo(filePath, format) {
        const all = this.getAll();
        let content;
        if (format === 'json') {
            content = JSON.stringify(all, null, 2);
        }
        else {
            content = all.map((r) => {
                const parts = [
                    `[${r.timestamp}]`,
                    r.key,
                    `${r.errorClass}: ${r.message}`,
                    `at ${r.filePath}:${r.line}`,
                ];
                if (r.count > 1)
                    parts.push(`×${r.count}`);
                if (r.context)
                    parts.push(`ctx:${r.context}`);
                return parts.join('  ');
            }).join('\n') + '\n';
        }
        const dir = path.dirname(filePath);
        if (dir && !fs.existsSync(dir))
            fs.mkdirSync(dir, { recursive: true });
        fs.writeFileSync(filePath, content, 'utf8');
    }
    getFilePath() { return this.filePath; }
}
