// amend/src/registry/Registry.ts
import * as fs   from 'fs';
import * as path from 'path';
import { decode as engineDecode, DecodedKey } from './CoreEngine.js';

// ── Types ─────────────────────────────────────────────────────────────

/** Raw line stored on disk — lean, no message/location. */
interface RawLine {
  key:       string;
  timestamp: string;
  context?:  string;
  metadata?: Record<string, unknown>[];
}

/** Consolidated record returned to callers. */
export interface ErrorRecord {
  key:        string;
  errorClass: string;
  message:    string;
  filePath:   string;
  line:       number;
  funcName:   string;
  timestamp:  string;
  count:      number;
  context?:   string;
  metadata:   Record<string, unknown>[];
}

export interface RegistryStatus {
  total:  number;
  recent: ErrorRecord[];
}

// ── Helpers ───────────────────────────────────────────────────────────

function safeDecodeKey(key: string): Partial<DecodedKey> {
  try { return engineDecode(key); } catch { return {}; }
}

function mergeToRecord(raw: RawLine): ErrorRecord {
  const d = safeDecodeKey(raw.key);
  return {
    key:        raw.key,
    errorClass: d.errorClass ?? '',
    message:    d.message    ?? '',
    filePath:   d.filePath   ?? '',
    line:       d.line       ?? 0,
    funcName:   d.funcName   ?? '',
    timestamp:  raw.timestamp,
    count:      1,
    context:    raw.context,
    metadata:   raw.metadata ?? [],
  };
}

// ── Registry ──────────────────────────────────────────────────────────

export class Registry {
  private static instance: Registry;
  private readonly dir:      string;
  private readonly filePath: string;

  private constructor() {
    this.dir      = path.join(process.cwd(), '.amend');
    this.filePath = path.join(this.dir, 'errors.log');
    this.ensureDir();
  }

  static getInstance(): Registry {
    if (!Registry.instance) Registry.instance = new Registry();
    return Registry.instance;
  }

  private ensureDir(): void {
    if (!fs.existsSync(this.dir)) fs.mkdirSync(this.dir, { recursive: true });
  }

  // ── Write ──────────────────────────────────────────────────────

  record(error: {
    key:       string;
    timestamp: string;
    context?:  string;
    metadata?: Record<string, unknown>[];
  }): void {
    const entry: RawLine = {
      key:       error.key,
      timestamp: error.timestamp,
      ...(error.context  ? { context:  error.context  } : {}),
      ...(error.metadata?.length ? { metadata: error.metadata } : {}),
    };
    fs.appendFileSync(this.filePath, JSON.stringify(entry) + '\n', 'utf8');
  }

  // ── Read ───────────────────────────────────────────────────────

  /**
   * Read all raw lines and consolidate by key:
   *   count    = number of occurrences
   *   metadata = all metadata arrays merged in order
   *   timestamp = most recent occurrence
   */
  getAll(): ErrorRecord[] {
    if (!fs.existsSync(this.filePath)) return [];
    const content = fs.readFileSync(this.filePath, 'utf8').trim();
    if (!content) return [];

    const index = new Map<string, ErrorRecord>();

    for (const rawLine of content.split('\n')) {
      if (!rawLine) continue;
      let raw: RawLine;
      try { raw = JSON.parse(rawLine) as RawLine; } catch { continue; }
      if (!raw.key) continue;

      const existing = index.get(raw.key);
      if (existing) {
        existing.count++;
        existing.timestamp = raw.timestamp; // keep latest
        if (raw.metadata?.length) existing.metadata.push(...raw.metadata);
      } else {
        index.set(raw.key, mergeToRecord(raw));
      }
    }

    return [...index.values()];
  }

  find(key: string): ErrorRecord[] {
    return this.getAll().filter((r) => r.key === key || r.key.startsWith(key));
  }

  clear(): void {
    if (fs.existsSync(this.filePath)) fs.writeFileSync(this.filePath, '', 'utf8');
  }

  status(): RegistryStatus {
    const all = this.getAll();
    return { total: all.length, recent: all.slice(-5) };
  }

  exportTo(filePath: string, format: 'json' | 'log'): void {
    const all = this.getAll();
    let content: string;

    if (format === 'json') {
      content = JSON.stringify(all, null, 2);
    } else {
      content = all.map((r) => {
        const parts = [
          `[${r.timestamp}]`,
          r.key,
          `${r.errorClass}: ${r.message}`,
          `at ${r.filePath}:${r.line}`,
        ];
        if (r.count > 1) parts.push(`×${r.count}`);
        if (r.context)   parts.push(`ctx:${r.context}`);
        return parts.join('  ');
      }).join('\n') + '\n';
    }

    const dir = path.dirname(filePath);
    if (dir && !fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
    fs.writeFileSync(filePath, content, 'utf8');
  }

  getFilePath(): string { return this.filePath; }
}
