// amend/dist/cjs/tests/logger.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.logger = void 0;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
class TestLogger {
    logFile;
    constructor() {
        this.logFile = path.join(process.cwd(), 'test-execution.log');
        // Clear out the previous run
        if (fs.existsSync(this.logFile)) {
            fs.writeFileSync(this.logFile, '', 'utf8');
        }
    }
    write(level, context, message) {
        const time = new Date().toISOString();
        const formatted = `[${time}] [${level}] [${context}] ${message}`;
        // Output to console with basic terminal colors
        if (level === 'ERROR') {
            console.error(`\x1b[31m${formatted}\x1b[0m`);
        }
        else if (level === 'WARN') {
            console.warn(`\x1b[33m${formatted}\x1b[0m`);
        }
        else if (level === 'SUCCESS') {
            console.log(`\x1b[32m${formatted}\x1b[0m`);
        }
        else {
            console.log(`\x1b[36m${formatted}\x1b[0m`); // cyan for INFO
        }
        // Append to file
        fs.appendFileSync(this.logFile, formatted + '\n', 'utf8');
    }
    info(context, message) {
        this.write('INFO', context, message);
    }
    warn(context, message) {
        this.write('WARN', context, message);
    }
    error(context, message) {
        this.write('ERROR', context, message);
    }
    success(context, message) {
        this.write('SUCCESS', context, message);
    }
}
exports.logger = new TestLogger();
