// amend/dist/esm/index.js
import { ScopeError } from './errors/scopeError.js';
import { NetworkError } from './errors/networkError.js';
import { ValidationError } from './errors/validationError.js';
import { Registry } from './registry/Registry.js';
import { CoreEngine } from './registry/CoreEngine.js';
import { AgentConfig } from './ai/AgentConfig.js';
const registry = Registry.getInstance();
const engine = CoreEngine;
const ai = AgentConfig.getInstance();
const amend = {
    ScopeError,
    NetworkError,
    ValidationError,
    engine: {
        encode: (...args) => CoreEngine.encode(...args),
        decode: (key) => CoreEngine.decode(key),
        verify: (key) => CoreEngine.verify(key),
        errorClassFrom: (key) => CoreEngine.errorClassFrom(key),
        locationFrom: (key) => CoreEngine.locationFrom(key),
        messageFrom: (key) => CoreEngine.messageFrom(key),
    },
    registry: {
        getAll: () => registry.getAll(),
        find: (key) => registry.find(key),
        clear: () => registry.clear(),
        status: () => registry.status(),
        exportTo: (filePath, format) => registry.exportTo(filePath, format),
    },
    ai: {
        enable: (opts) => ai.enable(opts.scope),
        disable: (opts) => ai.disable(opts.scope),
        hints: {
            set: (key, hint) => ai.setHint(key, hint),
            get: (key) => ai.getHint(key),
            clear: (key) => ai.clearHint(key),
        },
    },
};
export default amend;
export { ScopeError, NetworkError, ValidationError };
export { Registry } from './registry/Registry.js';
export { CoreEngine } from './registry/CoreEngine.js';
export { AgentConfig } from './ai/AgentConfig.js';
