// amend/dist/cjs/ai/AgentConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AgentConfig = void 0;
class AgentConfig {
    static instance;
    enabledScopes = new Set();
    disabledScopes = new Set();
    hintMap = new Map();
    constructor() { }
    static getInstance() {
        if (!AgentConfig.instance) {
            AgentConfig.instance = new AgentConfig();
        }
        return AgentConfig.instance;
    }
    enable(scope) {
        this.enabledScopes.add(scope);
        this.disabledScopes.delete(scope);
    }
    disable(scope) {
        this.disabledScopes.add(scope);
        this.enabledScopes.delete(scope);
    }
    isEnabled(filePath) {
        for (const scope of this.disabledScopes) {
            if (filePath.startsWith(scope))
                return false;
        }
        if (this.enabledScopes.size === 0)
            return true;
        for (const scope of this.enabledScopes) {
            if (filePath.startsWith(scope))
                return true;
        }
        return false;
    }
    getEnabledScopes() {
        return [...this.enabledScopes];
    }
    getDisabledScopes() {
        return [...this.disabledScopes];
    }
    setHint(hash, hint) {
        this.hintMap.set(hash, hint);
    }
    getHint(hash) {
        return this.hintMap.get(hash);
    }
    clearHint(hash) {
        this.hintMap.delete(hash);
    }
    getAllHints() {
        return Object.fromEntries(this.hintMap);
    }
}
exports.AgentConfig = AgentConfig;
