// amend/src/index.ts
import { ScopeError }      from './errors/scopeError.js';
import { NetworkError }     from './errors/networkError.js';
import { ValidationError }  from './errors/validationError.js';
import { Registry }         from './registry/Registry.js';
import { CoreEngine }       from './registry/CoreEngine.js';
import { AgentConfig }      from './ai/AgentConfig.js';

const registry = Registry.getInstance();
const engine   = CoreEngine;
const ai       = AgentConfig.getInstance();

const amend = {
  ScopeError,
  NetworkError,
  ValidationError,

  engine: {
    encode:        (...args: Parameters<typeof CoreEngine.encode>)        => CoreEngine.encode(...args),
    decode:        (key: string)                                          => CoreEngine.decode(key),
    verify:        (key: string)                                          => CoreEngine.verify(key),
    errorClassFrom:(key: string)                                          => CoreEngine.errorClassFrom(key),
    locationFrom:  (key: string)                                          => CoreEngine.locationFrom(key),
    messageFrom:   (key: string)                                          => CoreEngine.messageFrom(key),
  },

  registry: {
    getAll:   ()                                          => registry.getAll(),
    find:     (key: string)                               => registry.find(key),
    clear:    ()                                          => registry.clear(),
    status:   ()                                          => registry.status(),
    exportTo: (filePath: string, format: 'json' | 'log') => registry.exportTo(filePath, format),
  },

  ai: {
    enable:  (opts: { scope: string })    => ai.enable(opts.scope),
    disable: (opts: { scope: string })    => ai.disable(opts.scope),
    hints: {
      set:   (key: string, hint: string) => ai.setHint(key, hint),
      get:   (key: string)               => ai.getHint(key),
      clear: (key: string)               => ai.clearHint(key),
    },
  },
};

export default amend;
export { ScopeError, NetworkError, ValidationError };
export { Registry }    from './registry/Registry.js';
export { CoreEngine }  from './registry/CoreEngine.js';
export { AgentConfig } from './ai/AgentConfig.js';
export type { DecodedKey } from './registry/CoreEngine.js';
